sphinx/themes/basic/static/searchtools.js

```diff
@@ -466,14 +466,18 @@ const Search = {
       // add support for partial matches
       if (word.length > 2) {
         const escapedWord = _escapeRegExp(word);
-        Object.keys(terms).forEach((term) => {
-          if (term.match(escapedWord) && !terms[word])
-            arr.push({ files: terms[term], score: Scorer.partialTerm });
-        });
-        Object.keys(titleTerms).forEach((term) => {
-          if (term.match(escapedWord) && !titleTerms[word])
-            arr.push({ files: titleTerms[word], score: Scorer.partialTitle });
-        });
+        if (!terms.hasOwnProperty(word)) {
```

Comment:
Unless the performance improvement is significant, I think it'd be better to leave the conditional check after the `&&` clause.

(that's based on considering that if the first condition fails, then the clause after the `&&` isn't evaluated.  if moving the property-check eliminates a lot of regex matches, then fair enough, but this seems intended to be primarily a correctness fixup, so I'd lean towards making the logical diff as small as possible)

Reply:
I'm going to push back on this a bit: I think this small improvement is in scope (since we're changing the logic anyway) and the diff here is relatively small and easy to understand. Moving the property check should indeed reduce the number of regex matches a fair bit.

Reply:
Ok, no problem.  I think we'll have to agree to disagree on that.

Reply:
I'd prefer if you could avoid the extraction of an outer `if` statement here.  When viewing the diff, the larger set of line changes obscure what the bugfix is; and that's based on my assessment today, after having understood the fix a couple of weeks ago.  It'd be trickier for someone who wasn't familiar with the original pull request, or looking at it again after a longer duration of time.

Reply:
I stand by my original point. If you've convinced yourself that the code works, can we not just move on? We're talking about a <20 line diff here.

Reply:
I spend a decent amount of time reading old code changes and commit history, often related to small changes or fixes like this one, and it's usually much easier and less time-consuming to reason about minimal changes (with helpful commit messages ideally).

Narrowing the size of the diff _generally_ allows for increased review/inspection focus under those circumstances.  In this case the _behavioural_ diff isn't huge, but the code diff is larger than required to solve the bug.

In other words, when authoring, minimizing diffs can be a way to demonstrate that you value the time and attention of other developers (even ones who you may never meet or who may not exist yet!).

Reply:
Yes, I understood your point, I just don't agree with it in this case. I have more to say on this, but I think we've gone out of scope of a review. If you want to talk about it more, feel free to email me.

Reply:
Thanks for explaining.  From my perspective, bug report, code review and related discussion are often useful information when trying to understand a commit/change, so I'd prefer to keep the discussion public; in this case I do think that the size of the line diff is relevant to code review.

Would you be OK with me drafting an alternative PR?  I don't know if it'd be more likely to be accepted for merge, and it's OK if it isn't, but I'd like to present it as an alternative.

Reply:
> Thanks for explaining. From my perspective, bug report, code review and related discussion are often useful information when trying to understand a commit/change, so I'd prefer to keep the discussion public; in this case I do think that the size of the line diff is relevant to code review.

I don't think I have more to say on the technical front. But just to state it one more time so it's clear: I think this (small) PR as it stands fixes the issue in a more fundamental way and improves performance so it's my strong recommendation that it be the final state of the code.  I don't feel the effort of breaking it up into two PRs is worth the hassle and churn (leaving aside that this discussion has now occupied more effort than that would entail...).

> Would you be OK with me drafting an alternative PR? I don't know if it'd be more likely to be accepted for merge, and it's OK if it isn't, but I'd like to present it as an alternative.

It's not clear to me when someone who can actually merge the code will look at these changes. I can't stop you from proposing an alternative, but I think the best thing is probably just to wait for a maintainer to break our impasse. I am willing to make the change you suggest if it will actually get this merged and my PR is open to updates from the owners of this repository.

Reply:
Ok, after finding more time to understand the code, I concede that the performance improvements here are important.  I've opened #12045 with a description of what I believe the problem is.

Also, compared to my suggested alternative #12040, I think that the use of the JavaScript `hasOwnProperty` function is probably better/safer code style, making this pull request preferable.

@wlach could you edit the description of this pull request (#11958) to add 'Resolves #12045'?

(and thanks for your patience!)

Reply:
I wouldn't bother much about this. Here, it's more important to use `hasOwnProperty` instead of `!terms[word]` which could lead to incorrect boolean evaluations (and personally if we can have an easy-to-write improvement because we can isolate a condition from outside a loop, it's fine for me).

Reply:
I wouldn't bother much about this. Here, it's more important to use `hasOwnProperty` instead of `!terms[word]` which could lead to incorrect boolean evaluations (and personally if we can have an easy-to-write improvement because we can isolate a condition from outside a loop, it's fine for me).